compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/descriptors/IrDelegateDescriptor.kt

```diff
@@ -104,25 +104,19 @@ class IrPropertyDelegateDescriptorImpl(
 class IrImplementingDelegateDescriptorImpl(
     containingDeclaration: ClassDescriptor,
     delegateType: KotlinType,
-    override val correspondingSuperType: KotlinType
+    override val correspondingSuperType: KotlinType,
+    number: Int
 ) :
     IrDelegateDescriptorBase(
         containingDeclaration,
-        getDelegateName(containingDeclaration, correspondingSuperType),
+        Name.identifier("\$\$delegate_$number"),
         delegateType
     ),
     IrImplementingDelegateDescriptor
 
 internal fun getDelegateName(name: Name): Name =
     Name.identifier(name.asString() + "\$delegate")
 
-internal fun getDelegateName(classDescriptor: ClassDescriptor, superType: KotlinType): Name =
-    Name.identifier(
-        classDescriptor.name.asString() + "\$" +
```

Comment:
When I was working on this, I simply remove the containing class name, which could be unnamed. It was providing no useful information anyway since it's always the same class. The supertype name at least provided some information. Adding the index is still safer though since you could have multiple implemented interfaces with the same name.

Reply:
Thanks @punzki. Yes, I definitely want to add the number to disambiguate. I'm not sure we gain much from using different names than the current backend for these. The current backend uses $$delegate_n which is what this change ends up doing as well. @dnpetrov would you prefer that we keep the supertype name here if it exists?